File: frontend/src/utils/lookup.ts
import { syntheticAnalysis } from '../services/syntheticAnalysis';
import type {
  ComplaintSummary,
  CustomerLookupResponse,
  LookupRecord,
  TicketHistoryEvent,
  TicketRecord,
} from '../store';

function bySubmittedDesc<T extends { submitted_at?: string | null }>(left: T, right: T) {
  return String(right.submitted_at ?? '').localeCompare(String(left.submitted_at ?? ''));
}

export function buildLookupRecordsFromSummaries(complaints: ComplaintSummary[]): LookupRecord[] {
  return [...complaints]
    .sort(bySubmittedDesc)
    .map((summary) => {
      const detail = syntheticAnalysis(summary);
      const customer = detail.customer_profile!;
      const ticket = detail.ticket!;

      return {
        customer_id: customer.customer_id,
        full_name: customer.full_name,
        state: customer.state,
        credit_score: customer.credit_score,
        default_probability: customer.default_probability,
        previous_complaints_count: customer.previous_complaints_count,
        complaint_id: summary.complaint_id,
        ticket_id: ticket.ticket_id,
        ticket_status: ticket.status,
        product: summary.product ?? detail.classification?.product ?? 'Unknown',
        issue: summary.issue ?? detail.classification?.issue ?? 'General handling',
        risk_level: summary.risk_level ?? detail.compliance_risk?.risk_level ?? 'MEDIUM',
        criticality_score: summary.criticality_score ?? detail.criticality?.score ?? 0,
        assigned_team: summary.assigned_team ?? detail.routing?.assigned_team ?? 'Unassigned',
        queue: detail.internal_teams?.primary_team.queue ?? 'Complaint Operations',
        submitted_at: summary.submitted_at,
      };
    });
}

export function filterLookupRecords(records: LookupRecord[], query: string): LookupRecord[] {
  const needle = query.trim().toLowerCase();
  if (!needle) return records;

  return records.filter((record) =>
    [
      record.customer_id,
      record.full_name,
      record.complaint_id,
      record.ticket_id,
      record.product,
      record.issue,
      record.assigned_team,
      record.state,
    ]
      .filter(Boolean)
      .some((value) => String(value).toLowerCase().includes(needle))
  );
}

function mergeTimeline(tickets: TicketRecord[]): TicketHistoryEvent[] {
  return tickets
    .flatMap((ticket) => ticket.history ?? [])
    .sort((left, right) => String(right.timestamp ?? '').localeCompare(String(left.timestamp ?? '')))
    .slice(0, 24);
}

export function buildCustomerLookupFromSummaries(
  complaints: ComplaintSummary[],
  customerId: string,
): CustomerLookupResponse | null {
  const details = complaints
    .map((summary) => syntheticAnalysis(summary))
    .filter((detail) => detail.customer_profile?.customer_id === customerId)
    .sort((left, right) => String(right.submitted_at ?? '').localeCompare(String(left.submitted_at ?? '')));

  if (!details.length) return null;

  const latest = details[0];
  const profile = { ...latest.customer_profile! };
  const tickets = details.map((detail) => detail.ticket!).sort((left, right) => String(right.created_at ?? '').localeCompare(String(left.created_at ?? '')));
  const history = details.map((detail) => ({
    complaint_id: detail.complaint_id,
    ticket_id: detail.ticket?.ticket_id ?? '',
    product: detail.classification?.product ?? detail.complaint.product ?? 'Unknown',
    issue: detail.classification?.issue ?? 'General handling',
    risk_level: detail.compliance_risk?.risk_level ?? 'MEDIUM',
    criticality_score: detail.criticality?.score ?? 0,
    assigned_team: detail.routing?.assigned_team ?? 'Unassigned',
    status: detail.status,
    submitted_at: detail.submitted_at,
  }));

  profile.previous_complaints_count = Math.max(0, history.length - 1);

  return {
    customer_id: customerId,
    profile,
    metrics: {
      total_complaints: history.length,
      open_tickets: tickets.filter((ticket) => !['closed', 'resolved'].includes(ticket.status)).length,
      critical_cases: history.filter((item) => item.risk_level === 'CRITICAL').length,
      high_risk_cases: history.filter((item) => ['HIGH', 'CRITICAL'].includes(item.risk_level)).length,
      total_products: profile.open_products.length,
      total_loans: profile.open_products.filter((product) => product.toLowerCase().includes('loan') || product.toLowerCase().includes('mortgage')).length,
    },
    complaints: history,
    tickets,
    timeline: mergeTimeline(tickets),
    latest_complaint_id: latest.complaint_id,
  };
}

File: backend/services/lookup.py
"""Customer and ticket lookup helpers."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from backend.services.company_logic import enrich_detail


def _sort_details(details: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    return sorted(details, key=lambda detail: detail.get("submitted_at") or "", reverse=True)


def _ensure_enriched(detail: Dict[str, Any]) -> Dict[str, Any]:
    if detail.get("customer_profile") and detail.get("ticket"):
        return detail
    return enrich_detail(detail)


def _lookup_record(detail: Dict[str, Any]) -> Dict[str, Any]:
    enriched = _ensure_enriched(detail)
    complaint = enriched.get("complaint") or {}
    classification = enriched.get("classification") or {}
    compliance = enriched.get("compliance_risk") or {}
    customer = enriched.get("customer_profile") or {}
    ticket = enriched.get("ticket") or {}
    teams = enriched.get("internal_teams") or {}
    primary_team = teams.get("primary_team") or {}

    return {
        "customer_id": customer.get("customer_id"),
        "full_name": customer.get("full_name"),
        "state": customer.get("state") or complaint.get("customer_state"),
        "credit_score": customer.get("credit_score"),
        "default_probability": customer.get("default_probability"),
        "previous_complaints_count": customer.get("previous_complaints_count"),
        "complaint_id": enriched["complaint_id"],
        "ticket_id": ticket.get("ticket_id"),
        "ticket_status": ticket.get("status"),
        "product": classification.get("product") or complaint.get("product"),
        "issue": classification.get("issue"),
        "risk_level": compliance.get("risk_level"),
        "criticality_score": (enriched.get("criticality") or {}).get("score"),
        "assigned_team": primary_team.get("team_name") or (enriched.get("routing") or {}).get("assigned_team"),
        "queue": primary_team.get("queue"),
        "submitted_at": enriched.get("submitted_at"),
    }


def list_lookup_records(details: List[Dict[str, Any]], query: Optional[str] = None, limit: int = 100, offset: int = 0) -> Dict[str, Any]:
    records = [_lookup_record(detail) for detail in _sort_details(details)]
    needle = (query or "").strip().lower()

    if needle:
        records = [
            record
            for record in records
            if any(
                needle in str(value).lower()
                for value in (
                    record.get("customer_id"),
                    record.get("full_name"),
                    record.get("complaint_id"),
                    record.get("ticket_id"),
                    record.get("product"),
                    record.get("issue"),
                    record.get("assigned_team"),
                    record.get("state"),
                )
                if value
            )
        ]

    return {
        "records": records[offset:offset + limit],
        "total": len(records),
    }


def get_customer_lookup(details: List[Dict[str, Any]], customer_id: str) -> Optional[Dict[str, Any]]:
    enriched = [_ensure_enriched(detail) for detail in _sort_details(details)]
    matches = [detail for detail in enriched if (detail.get("customer_profile") or {}).get("customer_id") == customer_id]
    if not matches:
        return None

    latest = matches[0]
    profile = dict(latest.get("customer_profile") or {})
    profile["previous_complaints_count"] = max(0, len(matches) - 1)

    complaints = []
    tickets = []
    timeline = []

    for detail in matches:
        classification = detail.get("classification") or {}
        compliance = detail.get("compliance_risk") or {}
        routing = detail.get("routing") or {}
        complaint = detail.get("complaint") or {}
        ticket = detail.get("ticket") or {}

        complaints.append(
            {
                "complaint_id": detail["complaint_id"],
                "ticket_id": ticket.get("ticket_id"),
                "product": classification.get("product") or complaint.get("product"),
                "issue": classification.get("issue"),
                "risk_level": compliance.get("risk_level"),
                "criticality_score": (detail.get("criticality") or {}).get("score"),
                "assigned_team": routing.get("assigned_team"),
                "status": detail.get("status"),
                "submitted_at": detail.get("submitted_at"),
            }
        )
        tickets.append(ticket)
        timeline.extend(ticket.get("history") or [])

    timeline.sort(key=lambda entry: entry.get("timestamp") or "", reverse=True)
    complaints.sort(key=lambda item: item.get("submitted_at") or "", reverse=True)
    tickets.sort(key=lambda item: item.get("created_at") or "", reverse=True)

    open_products = profile.get("open_products") or []
    metrics = {
        "total_complaints": len(complaints),
        "open_tickets": sum(1 for ticket in tickets if ticket.get("status") not in {"closed", "resolved"}),
        "critical_cases": sum(1 for complaint in complaints if complaint.get("risk_level") == "CRITICAL"),
        "high_risk_cases": sum(1 for complaint in complaints if complaint.get("risk_level") in {"HIGH", "CRITICAL"}),
        "total_products": len(open_products),
        "total_loans": sum(1 for product in open_products if "loan" in str(product).lower() or "mortgage" in str(product).lower()),
    }

    return {
        "customer_id": customer_id,
        "profile": profile,
        "metrics": metrics,
        "complaints": complaints,
        "tickets": tickets,
        "timeline": timeline[:24],
        "latest_complaint_id": latest["complaint_id"],
    }
